lib/App/DuckPAN/Restart.pm

```diff
@@ -0,0 +1,98 @@
+package App::DuckPAN::Restart;
+# ABSTRACT: Automatic restarting of application on file change
+
+use File::Find::Rule;
+use Filesys::Notify::Simple;
+
+use strict;
+
+use Moo::Role;
+
+requires '_run_app';
+
+sub run_restarter {
+    my ($self, $args) = @_;
+
+    # will keep (re)starting the server until user ctrl-c
+    while(1){
+        defined(my $app = fork) or die 'Failed to fork application';
+        unless($app){ # app kid
+            $self->_run_app($args);
+            exit 0;
+        }
+
+        my $fmon = fork;
+        unless($fmon){ # file monitor kid
+            unless(defined $fmon){
+                kill SIGTERM => $app;
+                die 'Failed to fork file monitor';
+            }
+            $self->_monitor_directories;
+            exit 0;
+        }
+
+        # wait for one them to exit
+        my $pid = waitpid -1, 0;
+
+        # reload the application
+        if($pid == $fmon){
+            # if we can't kill the app, let's not start another
+            unless(kill SIGTERM => $app){
+                die "Failed to kill the application (pid: $app). Check manually";
+            }
+            # wait for it, otherwise the next whie loop will get it
+            waitpid($app, 0);
+        }
+        elsif($pid == $app){ # or exit
+            kill SIGTERM => $fmon;
+            exit;
+        }
+        else{ die "Unknown kid $pid reaped!\n"; } # shouldn't happen
+    }
+}
+
+# Monitors development directories for file changes.  Tries to get the
+# list of directories in a general way.  This subroutine 
+# blocks, so when it returns we know there's been a change
+sub _monitor_directories {
+    my $self  = shift;
+
+    # Find all of the directories that neeed to monitored
```

Comment:
add an issue or a comment here that we should improve this / make a function for it.